coreclient/src/clients/process/process_qs.rs

```diff
@@ -0,0 +1,456 @@
+        // loading AS credentials or fetching them from the AS.
+        let group = Group::join_group(
+            welcome_bundle,
+            &self.inner.key_store.wai_ear_key,
+            self.inner.connection.clone(),
+            &self.inner.api_clients,
+        )
+        .await?;
+        let group_id = group.group_id().clone();
+
+        // WelcomeBundle Phase 2: Store the user profiles of the group
+        // members if they don't exist yet and store the group and the
+        // new conversation.
+        let mut connection = self.inner.connection.lock().await;
+        let mut transaction = connection.transaction()?;
+        group
+            .members(&transaction)
+            .into_iter()
+            .try_for_each(|user_name| {
+                UserProfile::new(user_name, None, None).store(&transaction)
+            })?;
+
+        // Set the conversation attributes according to the group's
+        // group data.
+        let group_data = group.group_data().ok_or(anyhow!("No group data"))?;
+        let attributes: ConversationAttributes = PhnxCodec::from_slice(group_data.bytes())?;
+
+        let conversation = Conversation::new_group_conversation(group_id.clone(), attributes);
+        // If we've been in that conversation before, we delete the old
+        // conversation (and the corresponding MLS group) first and then
+        // create a new one. We do leave the messages intact, though.
+        Conversation::delete(&transaction, conversation.id())?;
+        Group::delete_from_db(&mut transaction, &group_id)?;
+        group.store(&transaction)?;
+        conversation.store(&transaction)?;
+        transaction.commit()?;
+        drop(connection);
+
+        Ok(ProcessQsMessageResult::NewConversation(conversation.id()))
+    }
+
+    async fn handle_mls_message(
+        &self,
+        mls_message: MlsMessageIn,
+        ds_timestamp: TimeStamp,
+    ) -> Result<ProcessQsMessageResult> {
+        let protocol_message: ProtocolMessage = match mls_message.extract() {
+            MlsMessageBodyIn::PublicMessage(handshake_message) =>
+                handshake_message.into(),
+            // Only application messages are private
+            MlsMessageBodyIn::PrivateMessage(app_msg) => app_msg.into(),
+            // Welcomes always come as a WelcomeBundle, not as an MLSMessage.
+            MlsMessageBodyIn::Welcome(_) |
+            // Neither GroupInfos nor KeyPackages should come from the queue.
+            MlsMessageBodyIn::GroupInfo(_) | MlsMessageBodyIn::KeyPackage(_) => bail!("Unexpected message type"),
+        };
+        // MLSMessage Phase 1: Load the conversation and the group.
+        let group_id = protocol_message.group_id();
+        let connection = self.inner.connection.lock().await;
+        let conversation = Conversation::load_by_group_id(&connection, group_id)?
+            .ok_or(anyhow!("No conversation found for group ID {:?}", group_id))?;
```

Comment:
minor: Most likely we want to use `ok_or_else` here, otherwise the anyhow error is always constructed.